src/crypto/src/id.rs

```diff
@@ -318,6 +323,265 @@ impl fmt::Display for DocumentId {
+                if input == i32::MIN {
+                    0
+                } else {
+                    let new_input = input as u32;
+                    (new_input & 0x7fffffff) as u32
+                }
+            }
+            false => {
+                let new_input = input as u32;
+                (new_input | 0x80000000) as u32
+            }
+        }
+    }
+    fn decode_u32(input: u32) -> i32 {
+        if input == 0 {
+            i32::MIN
+        } else {
+            if input as i32 > 0 {
+                (input | 0x80000000) as i32
+            } else {
+                (input & 0x7fffffff) as i32
+            }
+        }
+    }
+
+    fn encode_i64(input: i64) -> u64 {
+        match input < 0 {
+            true => {
+                if input == i64::MIN {
+                    0
+                } else {
+                    let new_input = input as u64;
+                    (new_input & 0x7fffffffffffffff) as u64
+                }
+            }
+            false => {
+                let new_input = input as u64;
+                (new_input | 0x8000000000000000) as u64
+            }
+        }
+    }
+    fn decode_u64(input: u64) -> i64 {
+        if input == 0 {
+            i64::MIN
+        } else {
+            if input as i64 > 0 {
+                (input | 0x8000000000000000) as i64
+            } else {
+                (input & 0x7fffffffffffffff) as i64
+            }
+        }
+    }
+
+    fn add(&mut self, field: &Option<Bson>) -> std::result::Result<(), DB3Error> {
+        match field {
+            None => {
+                self.data
+                    .write_u8(FieldTypeId::Null as u8)
+                    .map_err(|e| DB3Error::DocumentDecodeError(format!("{e}")))?;
+            }
+
+            Some(Bson::Boolean(b)) => {
+                self.data
+                    .write_u8(FieldTypeId::Bool as u8)
+                    .map_err(|e| DB3Error::DocumentDecodeError(format!("{e}")))?;
+                self.data
+                    .write_u8(*b as u8)
+                    .map_err(|e| DB3Error::DocumentDecodeError(format!("{e}")))?;
+            }
+            Some(Bson::Int64(n)) => {
+                self.data
+                    .write_u8(FieldTypeId::I64 as u8)
+                    .map_err(|e| DB3Error::DocumentDecodeError(format!("{e}")))?;
+                self.data
+                    .write_u64::<BigEndian>(Self::encode_i64(*n))
+                    .map_err(|e| DB3Error::DocumentDecodeError(format!("{e}")))?;
+            }
+            Some(Bson::Int32(n)) => {
+                self.data
+                    .write_u8(FieldTypeId::I32 as u8)
+                    .map_err(|e| DB3Error::DocumentDecodeError(format!("{e}")))?;
+                self.data
+                    .write_u32::<BigEndian>(Self::encode_i32(*n))
+                    .map_err(|e| DB3Error::DocumentDecodeError(format!("{e}")))?;
+            }
+            // TODO: add \0 as the end of string.
```

Comment:
Did the MongoDB use `\0` to recognize the end of the string?

Reply:
yep it seems like that. 
https://github.com/mongodb/mongo/blob/924f756576cfaad5a4253736394ae3e8973c79bb/src/mongo/db/storage/key_string.cpp#L1223-L1225